README.md
# Webcryptobox
Convenient wrappers around [OpenSSL](https://docs.rs/openssl/latest/openssl/) to use [WebCrypto](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) compatible cryptography.

It works nicely together with the [JavaScript Webcryptobox](https://github.com/jo/webcryptobox-js) and [Bash Webcryptobox](https://github.com/jo/webcryptobox-sh).

Webcryptobox is **both a library and a little command line interface.** It helps with elliptic curve key generation, derivation, fingerprinting, import and export as well as AES encryption and decryption.

For the library documentation, see [docs.rs/webcryptobox](https://docs.rs/webcryptobox/latest/webcryptobox/). In the following I describe the CLI usage.

## Installation
Clone the project, compile it and use the binary:

```sh
git clone https://github.com/jo/webcryptobox-js
cd webcryptobox-js
cargo build --release
sudo ln -s `pwd`/target/release/webcryptobox /usr/local/bin/webcryptobox
```

## Usage
Webcryptobox prints out usage information if you do not provide any argument, or if the command is not recognized.

### Configuration
Cipher selection is done via environment variables:

* `CURVE`: EC curve name. Defaults to `P-521`
* `MODE`: AES mode. Defaults to `GCM`
* `LENGTH`: AES key length in bits. Default is `256`

#### Supported EC Curves
* `P-256`: 256-bit prime field Weierstrass curve. Also known as `secp256r1` or `prime256v1`.
* `P-384`: 384-bit prime field Weierstrass curve. Also known as: `secp384r1` or `ansip384r1`.
* `P-521`: 521-bit prime field Weierstrass curve. Also known as: `secp521r1` or `ansip521r1`.

#### Supported AES Modes
* `CBC`: Cipher Block Chaining Mode
* `GCM`: Galois/Counter Mode

#### Supported AES Key Lengths
* `128`
* `256`

### `generate-key-pair`
Generate ECDH key and output private key as pem.

```sh
$ webcryptobox generate-key-pair
-----BEGIN PRIVATE KEY-----
MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIANqKw4HlT3vUqNQTg
YtueEyGDmnm8bOo50Gyq/Y7sSldOZzcPQ+WggGnGd62M6sYit3IZZ+/752OR7+8O
td1swiuhgYkDgYYABAGga0QN+HqG9QQwSACVUvts2AV8gbJ+iUKZJWWXQ5BcCoyX
RBVi21Ga0ViWQSXyump3H6YrdXq2+2QVKcpQMQmRZQB3P4K6+TqKX4RxXcLKdu0F
F7fvYfUXLAxh71MN6HUQDguJ2rwtiM9DfXBu/kqvMURPnb09Cv/ymXN6mU+iavJ9
3A==
-----END PRIVATE KEY-----
```

### `generate-key`
Generate AES key, output as hex.

```sh
$ webcryptobox generate-key
324cc75bdd8185e84c3f2ede3d67017f6fe22ad45dade188b8233dae0f836304
```

### `generate-iv`
Generate an initialization vector, output as hex.

```sh
$ webcryptobox generate-iv
a0e58789d6a4ffa9cec8bbd2
```

### `derive-public-key <filename>`
Derive public key from private key PEM file, output public key pem.

```sh
$ webcryptobox derive-public-key private-key.pem 
-----BEGIN PUBLIC KEY-----
MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQBoGtEDfh6hvUEMEgAlVL7bNgFfIGy
folCmSVll0OQXAqMl0QVYttRmtFYlkEl8rpqdx+mK3V6tvtkFSnKUDEJkWUAdz+C
uvk6il+EcV3CynbtBRe372H1FywMYe9TDeh1EA4Lidq8LYjPQ31wbv5KrzFET529
PQr/8plzeplPomryfdw=
-----END PUBLIC KEY-----
```

### `sha1-fingerprint <filename>`
Calculate SHA-1 fingerprint from private or public key pem file, output hex. Works by hashing the DER contents of the public key part.

```sh
$ webcryptobox sha1-fingerprint private-key.pem 
e635c9976397b68fcc5a3e872fad6730c86e945b
```

### `sha256-fingerprint <filename>`
Calculate SHA-256 fingerprint from private or public key pem file, output hex. Works by hashing the DER contents of the public key part.

```sh
$ webcryptobox sha256-fingerprint private-key.pem 
bfeb4a6ea6c4be9117800516ebb4ca9527f106a3498e3684ee4dc291470752aa
```

### `derive-key <private-key> <peer-key>`
derive AES key from private and public key files as pem, output hex.

```sh
$ webcryptobox derive-key private-key.pem their-public-key.pem 
0028ca6b01e9b6293692217efe690bcc2e6537e18cbf3cf7e238adf99f2fe392
```

### `encrypt <key> <iv>`
reads message from STDIN and key and iv as hex args, encrypts message and prints out as base64.

```sh
$ echo "a secret message" \
  | webcryptobox encrypt \
    324cc75bdd8185e84c3f2ede3d67017f6fe22ad45dade188b8233dae0f836304 \
    a0e58789d6a4ffa9cec8bbd2
yMQpXTpEkXn9qcav4YJnudO3OB73ywKRnKjzNhckh3Pw
```

### `decrypt <key> <iv>`
reads encrypted message in base64 format from STDIN and key and iv as hex args and prints out decrypted message.

```sh
$ echo -n "yMQpXTpEkXn9qcav4YJnudO3OB73ywKRnKjzNhckh3Pw" \
  | webcryptobox decrypt \
    324cc75bdd8185e84c3f2ede3d67017f6fe22ad45dade188b8233dae0f836304 \
    a0e58789d6a4ffa9cec8bbd2
a secret message
```

### `derive-and-encrypt <private-key> <peer-key> <iv>`
reads message from STDIN and private and public key files as pem and iv as hex and encrypts message and prints out as base64.

```sh
$ echo "a secret message" \
  | webcryptobox derive-and-encrypt \
    my-private-key.pem \
    their-public-key.pem \
    b3b7013cdd605017fc1a4010
fBAXWxsdrctjXyceuuAfZMFWb6hLjFGOZAngYlbDTaq6
```

### ` derive-and-decrypt <private-key> <peer-key> <iv>`
reads encrypted message in base64 format from STDIN, private and public key files as pem and iv as hex and prints out decrypted message.

```sh
$ echo -n "fBAXWxsdrctjXyceuuAfZMFWb6hLjFGOZAngYlbDTaq6" \
  | webcryptobox derive-and-decrypt \
    their-private-key.pem \
    my-public-key.pem \
    b3b7013cdd605017fc1a4010
a secret message
```


## Test
There's a test suite which ensures the lib works as expected. Run it with cargo:
```sh
cargo test
```

## License
This package is licensed under the [Apache 2.0 License](https://www.apache.org/licenses/LICENSE-2.0).

© 2022 Johannes J. Schmidt


CHANGELOG
## v1.0.5 - More Examples
Adds a highlevel example

## v1.0.4 - Docs Improved II
Further improved documentation

## v1.0.3 - Docs Improved
More examples

## v1.0.2 - Doc Examples
Use executing examples

## v1.0.1 - Package Metadata
Add package metadata such as repository etc.

## v1.0.0 - Hello World!
Initial release of Webcryptobox


src/lib.rs
#![warn(missing_docs)]
#![warn(rustdoc::missing_doc_code_examples)]

//! Webcryptobox provides convenient wrappers around
//! [OpenSSL](https://docs.rs/openssl/latest/openssl/) to use
//! [WebCrypto](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) compatible
//! cryptography.
//!
//! It works nicely together with the [JavaScript
//! Webcryptobox](https://github.com/jo/webcryptobox-js) and [Bash
//! Webcryptobox](https://github.com/jo/webcryptobox-sh).
//!
//! Webcryptobox helps with elliptic curve key generation, derivation, fingerprinting, import and
//! export as well as AES encryption and decryption.
//!
//! # Example:
//!
//! ```rust
//! let wcb = webcryptobox::Webcryptobox::default();
//!
//! // Alice creates a key and sends her public key pem to Bob
//! let alice = wcb.generate_key_pair().unwrap();
//! let alice_public_key = wcb.derive_public_key(&alice).unwrap();
//! let alice_public_key_pem = wcb.export_public_key_pem(&alice_public_key).unwrap();
//!
//! // Bob also creates a key and sends his public key pem to Alice
//! let bob = wcb.generate_key_pair().unwrap();
//! let bobs_public_key = wcb.derive_public_key(&bob).unwrap();
//! let bob_public_key_pem = wcb.export_public_key_pem(&bobs_public_key).unwrap();
//!
//! // Alice uses Bobs public key to derive a shared key
//! let bobs_key = wcb.import_public_key_pem(&bob_public_key_pem).unwrap();
//! let alice_shared_key = wcb.derive_key(alice, bobs_key).unwrap();
//!
//! // She now encrypts a message and sends the encrypted message and the iv to Bob
//! let iv = wcb.generate_iv().unwrap();
//! let data = (b"a secret message").to_vec();
//! let encrypted_message = wcb.encrypt(&alice_shared_key, &iv, &data).unwrap();
//!
//! // Now Bob derives the same shared secret
//! let alice_key = wcb.import_public_key_pem(&alice_public_key_pem).unwrap();
//! let bobs_shared_key = wcb.derive_key(bob, alice_key).unwrap();
//!
//! // and decrypts the message
//! let message = wcb.decrypt(&bobs_shared_key, &iv, &encrypted_message);
//! ```

use openssl::derive::Deriver;
use openssl::ec::{EcGroup, EcKey};
use openssl::error::ErrorStack;
use openssl::nid::Nid;
use openssl::pkey::PKey;
use openssl::pkey::{Private, Public};
use openssl::rand::rand_bytes;
use openssl::sha;
use openssl::symm;

// Length of the AEAD tag used in GCM mode
const AEAD_TAG_LENGTH: usize = 16;

/// Holds cipher configuration and provides the methods for dancing with elliptic curves and
/// advanced encryption.
///
/// Use the factories [`webcryptobox::Webcryptobox::new`] or
/// [`webcryptobox::Webcryptobox::default`] to instantiate a Webcryptobox:
///
/// # Example:
///
/// ```rust
/// // P-512, AES 256 GCM
/// let wcb_default = webcryptobox::Webcryptobox::default();
/// // same as
/// let wcb_p256_aes_256_gcm = webcryptobox::Webcryptobox::new("P-521", "GCM", 256);
/// // P-256, AES 128 CBC
/// let wcb_p256_aes_128_cbc = webcryptobox::Webcryptobox::new("P-256", "CBC", 128);
/// ```
pub struct Webcryptobox {
    is_aead: bool,
    group: EcGroup,
    cipher: symm::Cipher,
}

impl Webcryptobox {
    /// Creates a new Webcryptobox.
    ///
    /// # Supported Curves:
    /// * `P-256` [openssl::nid::Nid::X9_62_PRIME256V1`] aka `secp256r1` or `prime256v1`
    /// * `P-384` [openssl::nid::Nid::SECP384R1`] aka `secp384r1` or `ansip384r1`
    /// * `P-521` [openssl::nid::Nid::SECP521R1`] aka `secp521r1` or `ansip521r1`
    ///
    /// # Supported AES Modes:
    /// * `CBC`: Cipher Block Chaining Mode
    /// * `GCM`: Galois/Counter Mode
    ///
    /// # Supported AES key Lengths:
    /// * `128`
    /// * `256`
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::new("P-256", "CBC", 128);
    /// ```
    /// This example uses the curve `P-256`, AES in `CBC` mode with a key length of 128.
    pub fn new(curve: &str, mode: &str, length: usize) -> Webcryptobox {
        let curve = match curve {
            "P-256" => Nid::X9_62_PRIME256V1,
            "P-384" => Nid::SECP384R1,
            "P-521" => Nid::SECP521R1,

            _ => panic!("Unsupported curve '{}'", curve),
        };

        let cipher = match mode {
            "CBC" => match length {
                128 => symm::Cipher::aes_128_cbc(),
                256 => symm::Cipher::aes_256_cbc(),
                _ => panic!("Unsupported length '{}'", length),
            },
            "GCM" => match length {
                128 => symm::Cipher::aes_128_gcm(),
                256 => symm::Cipher::aes_256_gcm(),
                _ => panic!("Unsupported length '{}'", length),
            },

            _ => panic!("Unsupported mode '{}'", mode),
        };

        let group = EcGroup::from_curve_name(curve).unwrap();
        let is_aead = mode == "GCM";

        Webcryptobox {
            group,
            cipher,
            is_aead,
        }
    }

    /// Creates a Webcryptobox with defaults (`curve=P-521, mode=GCM, length=256`).
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// ```
    pub fn default() -> Webcryptobox {
        let curve = Nid::SECP521R1;
        let group = EcGroup::from_curve_name(curve).unwrap();
        let cipher = symm::Cipher::aes_256_gcm();

        Webcryptobox {
            is_aead: true,
            group,
            cipher,
        }
    }

    /// Generate an EC private key.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair();
    /// ```
    pub fn generate_key_pair(&self) -> Result<EcKey<Private>, ErrorStack> {
        EcKey::generate(&self.group)
    }

    /// Given a private EC key, derives the public EC key.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair().unwrap();
    /// let public_key = wcb.derive_public_key(&key);
    /// ```
    pub fn derive_public_key(
        &self,
        private_key: &EcKey<Private>,
    ) -> Result<EcKey<Public>, ErrorStack> {
        let key_point = private_key.public_key();

        EcKey::from_public_key(&self.group, &key_point)
    }

    /// Import a private key PEM.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let pem = (b"-----BEGIN PRIVATE KEY-----
    /// MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIBcf8zEjlssqn4aTEB
    /// RR43ofwH/4BAXDAAd83Kz1Dyd+Ko0pit4ESgqSu/bJMdnDrpiGYuz0Klarwip8LD
    /// rYd9mEahgYkDgYYABAF2Nu9XKPs2CVFocuqCfaX5FzDUt6/nT/3Evqq8jBhK/ziN
    /// TrEs4wkZjuei5TS25aabX6iMex3etoN/GOw1KYpI4QBtIUnWudG8FT8N+USHSL9G
    /// h9fi+Yofeq4Io9DxPU1ChCKPIoQ6ORAMWoOCk9bTdIy6yqx33+RIM04wub4QAgDo
    /// LQ==
    /// -----END PRIVATE KEY-----").to_vec();
    /// let key = wcb.import_private_key_pem(&pem);
    /// ```
    pub fn import_private_key_pem(&self, pem: &[u8]) -> Result<EcKey<Private>, ErrorStack> {
        EcKey::private_key_from_pem(pem)
    }

    /// Import a public key PEM.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let pem = (b"-----BEGIN PUBLIC KEY-----
    /// MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQBdjbvVyj7NglRaHLqgn2l+Rcw1Lev
    /// 50/9xL6qvIwYSv84jU6xLOMJGY7nouU0tuWmm1+ojHsd3raDfxjsNSmKSOEAbSFJ
    /// 1rnRvBU/DflEh0i/RofX4vmKH3quCKPQ8T1NQoQijyKEOjkQDFqDgpPW03SMusqs
    /// d9/kSDNOMLm+EAIA6C0=
    /// -----END PUBLIC KEY-----").to_vec();
    /// let key = wcb.import_public_key_pem(&pem);
    /// ```
    pub fn import_public_key_pem(&self, pem: &[u8]) -> Result<EcKey<Public>, ErrorStack> {
        EcKey::public_key_from_pem(pem)
    }

    /// Export a private EC key in PEM format.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair().unwrap();
    /// let pem = wcb.export_private_key_pem(key);
    /// ```
    pub fn export_private_key_pem(
        &self,
        private_key: EcKey<Private>,
    ) -> Result<Vec<u8>, ErrorStack> {
        let key = PKey::from_ec_key(private_key).unwrap();
        key.private_key_to_pem_pkcs8()
    }

    /// Export a public EC key in PEM format.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair().unwrap();
    /// let public_key = wcb.derive_public_key(&key).unwrap();
    /// let pem = wcb.export_public_key_pem(&public_key);
    /// ```
    pub fn export_public_key_pem(&self, public_key: &EcKey<Public>) -> Result<Vec<u8>, ErrorStack> {
        public_key.public_key_to_pem()
    }

    /// Calculate a SHA-1 fingeprint from a private key.
    ///
    /// This hashes the DER data of the public key part of the key. Note this does not create a
    /// hash of the private key, but the public key.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair().unwrap();
    /// let fingerprint = wcb.sha1_fingerprint_from_private_key(&key);
    /// ```
    pub fn sha1_fingerprint_from_private_key(
        &self,
        key: &EcKey<Private>,
    ) -> Result<String, ErrorStack> {
        let der = key.public_key_to_der().unwrap();
        let mut hasher = sha::Sha1::new();
        hasher.update(&der);
        let hash = hasher.finish();
        let hex = hex::encode(hash);

        Ok(hex)
    }

    /// Calculate a SHA-1 fingeprint from a public key.
    ///
    /// This hashes the DER data of the key.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair().unwrap();
    /// let public_key = wcb.derive_public_key(&key).unwrap();
    /// let fingerprint = wcb.sha1_fingerprint_from_public_key(&public_key);
    /// ```
    pub fn sha1_fingerprint_from_public_key(
        &self,
        key: &EcKey<Public>,
    ) -> Result<String, ErrorStack> {
        let der = key.public_key_to_der().unwrap();
        let mut hasher = sha::Sha1::new();
        hasher.update(&der);
        let hash = hasher.finish();
        let hex = hex::encode(hash);

        Ok(hex)
    }

    /// Calculate a SHA-256 fingeprint from a private key.
    ///
    /// This hashes the DER data of the public key part of the key. Note this does not create a
    /// hash of the private key, but the public key.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair().unwrap();
    /// let fingerprint = wcb.sha256_fingerprint_from_private_key(&key);
    /// ```
    pub fn sha256_fingerprint_from_private_key(
        &self,
        key: &EcKey<Private>,
    ) -> Result<String, ErrorStack> {
        let der = key.public_key_to_der().unwrap();
        let mut hasher = sha::Sha256::new();
        hasher.update(&der);
        let hash = hasher.finish();
        let hex = hex::encode(hash);

        Ok(hex)
    }

    /// Calculate a SHA-256 fingeprint from a public key.
    ///
    /// This hashes the DER data of the key.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key_pair().unwrap();
    /// let public_key = wcb.derive_public_key(&key).unwrap();
    /// let fingerprint = wcb.sha256_fingerprint_from_public_key(&public_key);
    /// ```
    pub fn sha256_fingerprint_from_public_key(
        &self,
        key: &EcKey<Public>,
    ) -> Result<String, ErrorStack> {
        let der = key.public_key_to_der().unwrap();
        let mut hasher = sha::Sha256::new();
        hasher.update(&der);
        let hash = hasher.finish();
        let hex = hex::encode(hash);

        Ok(hex)
    }

    /// Generate AES key material to be used with `encrypt` and `decrypt`.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key();
    /// ```
    pub fn generate_key(&self) -> Result<Vec<u8>, ErrorStack> {
        match self.cipher.key_len() {
            16 => {
                let mut key = [0; 16];
                rand_bytes(&mut key).unwrap();
                Ok(key.to_vec())
            }
            32 => {
                let mut key = [0; 32];
                rand_bytes(&mut key).unwrap();
                Ok(key.to_vec())
            }
            _ => panic!("unknown size"),
        }
    }

    /// Derives AES key material to be used with `encrypt` and `decrypt` from given private and
    /// public key.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let alice = wcb.generate_key_pair().unwrap();
    /// let bob = wcb.generate_key_pair().unwrap();
    /// let bobs_public_key = wcb.derive_public_key(&bob).unwrap();
    /// let key = wcb.derive_key(alice, bobs_public_key);
    /// ```
    pub fn derive_key(
        &self,
        private_key: EcKey<Private>,
        public_key: EcKey<Public>,
    ) -> Result<Vec<u8>, ErrorStack> {
        let private_key = PKey::from_ec_key(private_key).unwrap();
        let public_key = PKey::from_ec_key(public_key).unwrap();

        let mut deriver = Deriver::new(&private_key).unwrap();
        deriver.set_peer(&public_key).unwrap();

        match self.cipher.key_len() {
            16 => {
                let mut key: [u8; 16] = [0; 16];
                deriver.derive(&mut key).unwrap();
                Ok(key.to_vec())
            }
            32 => {
                let mut key: [u8; 32] = [0; 32];
                deriver.derive(&mut key).unwrap();
                Ok(key.to_vec())
            }
            _ => panic!("unknown key size"),
        }
    }

    /// Generate AES initialization vector to be used with `encrypt` and `decrypt`.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let iv = wcb.generate_iv();
    /// ```
    pub fn generate_iv(&self) -> Result<Vec<u8>, ErrorStack> {
        match self.cipher.iv_len() {
            Some(12) => {
                let mut key: [u8; 12] = [0; 12];
                rand_bytes(&mut key).unwrap();
                Ok(key.to_vec())
            }
            Some(16) => {
                let mut key: [u8; 16] = [0; 16];
                rand_bytes(&mut key).unwrap();
                Ok(key.to_vec())
            }
            _ => panic!("unknown iv size"),
        }
    }

    /// Encrypts data with key and iv.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    /// let key = wcb.generate_key().unwrap();
    /// let iv = wcb.generate_iv().unwrap();
    /// let data = (b"a secret message").to_vec();
    /// let encrypted_message = wcb.encrypt(&key, &iv, &data);
    /// ```
    pub fn encrypt(&self, key: &[u8], iv: &Vec<u8>, data: &[u8]) -> Result<Vec<u8>, ErrorStack> {
        match self.is_aead {
            true => {
                let mut tag: [u8; AEAD_TAG_LENGTH] = [0; AEAD_TAG_LENGTH];
                let mut ciphertext =
                    symm::encrypt_aead(self.cipher, key, Some(iv), &[], data, &mut tag).unwrap();
                ciphertext.extend(tag);
                Ok(ciphertext)
            }
            false => symm::encrypt(self.cipher, key, Some(iv), data),
        }
    }

    /// Decrypts data with key and iv.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    ///
    /// let key = wcb.generate_key().unwrap();
    /// let iv = wcb.generate_iv().unwrap();
    /// let data = (b"a secret message").to_vec();
    /// let encrypted_message = wcb.encrypt(&key, &iv, &data).unwrap();
    ///
    /// let message = wcb.decrypt(&key, &iv, &encrypted_message);
    /// ```
    pub fn decrypt(&self, key: &[u8], iv: &Vec<u8>, data: &[u8]) -> Result<Vec<u8>, ErrorStack> {
        match self.is_aead {
            true => {
                let tag_start = data.len() - AEAD_TAG_LENGTH;
                let d = &data[0..tag_start];
                let tag = &data[tag_start..];
                symm::decrypt_aead(self.cipher, key, Some(iv), &[], &d, &tag)
            }
            false => symm::decrypt(self.cipher, key, Some(iv), &data),
        }
    }

    /// Derives AES key from given private and public key and encrypts message.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    ///
    /// let alice = wcb.generate_key_pair().unwrap();
    /// let bob = wcb.generate_key_pair().unwrap();
    /// let bobs_public_key = wcb.derive_public_key(&bob).unwrap();
    ///
    /// let iv = wcb.generate_iv().unwrap();
    /// let data = (b"a secret message").to_vec();
    /// let encrypted_message = wcb.derive_and_encrypt(alice, bobs_public_key, &iv, &data);
    /// ```
    pub fn derive_and_encrypt(
        &self,
        private_key: EcKey<Private>,
        public_key: EcKey<Public>,
        iv: &Vec<u8>,
        data: &[u8],
    ) -> Result<Vec<u8>, ErrorStack> {
        let key = self.derive_key(private_key, public_key).unwrap();

        self.encrypt(&key, iv, data)
    }

    /// Derives AES key from given private and public key and decrypts message.
    ///
    /// # Example:
    ///
    /// ```rust
    /// let wcb = webcryptobox::Webcryptobox::default();
    ///
    /// let alice = wcb.generate_key_pair().unwrap();
    /// let bob = wcb.generate_key_pair().unwrap();
    /// let alice_public_key = wcb.derive_public_key(&alice).unwrap();
    /// let bobs_public_key = wcb.derive_public_key(&bob).unwrap();
    ///
    /// let iv = wcb.generate_iv().unwrap();
    /// let data = (b"a secret message").to_vec();
    /// let encrypted_message = wcb.derive_and_encrypt(alice, bobs_public_key, &iv, &data).unwrap();
    ///
    /// let message = wcb.derive_and_decrypt(bob, alice_public_key, &iv, &encrypted_message);
    /// ```
    pub fn derive_and_decrypt(
        &self,
        private_key: EcKey<Private>,
        public_key: EcKey<Public>,
        iv: &Vec<u8>,
        data: &[u8],
    ) -> Result<Vec<u8>, ErrorStack> {
        let key = self.derive_key(private_key, public_key).unwrap();

        self.decrypt(&key, iv, data)
    }
}


src/main.rs
use webcryptobox::Webcryptobox;

use std::env;
use std::fs;
use std::io;
use std::io::Read;

fn help() {
    println!("Usage: webcryptobox <command> [arguments]

generate-key-pair
    Generate ECDH key and output private key as pem
generate-key
    Generate AES key, output as hex
generate-iv
    Generate an initialization vector, output as hex
derive-public-key <filename>
    Derive public key from private key PEM file, output public key pem
sha1-fingerprint <filename>
    Calculate SHA-1 fingerprint from private or public key pem file, output hex
sha256-fingerprint <filename>
    Calculate SHA-256 fingerprint from private or public key pem file, output hex
derive-key <private-key-filename> <public-key-filename>
    derive AES key from private and public key files as pem, output hex
encrypt <key> <iv>
    reads message from STDIN and key and iv as hex args,
    encrypts message and prints out as base64
decrypt <key> <iv>
    reads encrypted message in base64 format from STDIN and key and iv as hex args
    prints out decrypted message
derive-and-encrypt <private-key-filename> <public-key-filename> <iv>
    reads message from STDIN and private and public key files as pem and iv as hex
    encrypts message and prints out as base64
derive-and-decrypt <private-key-filename> <public-key-filename> <iv>
    reads encrypted message in base64 format from STDIN, private and public key files as pem and iv as hex
    prints out decrypted message

Cipher can be configured via the following environment variables:
  CURVE:  can be P-521, P-384 or P-256. Defaults to P-521
  MODE:   can be GCM or CBC. Defaults to GCM
  LENGTH: can be 256 or 128. Default is 256");
}

pub fn main() {
    let curve = env::var("CURVE").unwrap_or("P-521".to_string());
    let mode = env::var("MODE").unwrap_or("GCM".to_string());
    let length = env::var("LENGTH").unwrap_or("256".to_string());
    let length = length.parse::<usize>().unwrap();

    let wcb = Webcryptobox::new(&curve, &mode, length);

    let args: Vec<String> = env::args().collect();

    match args.len() {
        1 => help(),
        2 => {
            let cmd = &args[1];
            match &cmd[..] {
                "generate-key-pair" => {
                    let key = wcb.generate_key_pair().unwrap();
                    let pem = wcb.export_private_key_pem(key).unwrap();
                    let pem_string = std::str::from_utf8(&pem).unwrap();
                    println!("{}", pem_string)
                }
                "generate-key" => {
                    let key = wcb.generate_key().unwrap();
                    let key_string = hex::encode(key);
                    println!("{}", key_string)
                }
                "generate-iv" => {
                    let key = wcb.generate_iv().unwrap();
                    let key_string = hex::encode(key);
                    println!("{}", key_string)
                }
                _ => help(),
            }
        }
        3 => {
            let cmd = &args[1];
            match &cmd[..] {
                "derive-public-key" => {
                    let filename = &args[2];

                    let pem = fs::read(filename).unwrap();
                    let key = wcb.import_private_key_pem(&pem).unwrap();
                    let public_key = wcb.derive_public_key(&key).unwrap();
                    let pem = wcb.export_public_key_pem(&public_key).unwrap();
                    let pem_string = std::str::from_utf8(&pem).unwrap();
                    println!("{}", pem_string)
                }
                "sha1-fingerprint" => {
                    let filename = &args[2];

                    let pem = fs::read(filename).unwrap();
                    let fingerprint = match pem.starts_with(b"-----BEGIN PRIVATE KEY-----") {
                        true => {
                            let key = wcb.import_private_key_pem(&pem).unwrap();
                            wcb.sha1_fingerprint_from_private_key(&key).unwrap()
                        }
                        _ => {
                            let key = wcb.import_public_key_pem(&pem).unwrap();
                            wcb.sha1_fingerprint_from_public_key(&key).unwrap()
                        }
                    };
                    println!("{}", fingerprint)
                }
                "sha256-fingerprint" => {
                    let filename = &args[2];

                    let pem = fs::read(filename).unwrap();
                    let fingerprint = match pem.starts_with(b"-----BEGIN PRIVATE KEY-----") {
                        true => {
                            let key = wcb.import_private_key_pem(&pem).unwrap();
                            wcb.sha256_fingerprint_from_private_key(&key).unwrap()
                        }
                        _ => {
                            let key = wcb.import_public_key_pem(&pem).unwrap();
                            wcb.sha256_fingerprint_from_public_key(&key).unwrap()
                        }
                    };
                    println!("{}", fingerprint)
                }
                _ => help(),
            }
        }
        4 => {
            let cmd = &args[1];
            match &cmd[..] {
                "derive-key" => {
                    let private_key_filename = &args[2];
                    let public_key_filename = &args[3];

                    let private_key_pem = fs::read(private_key_filename).unwrap();
                    let private_key = wcb.import_private_key_pem(&private_key_pem).unwrap();

                    let public_key_pem = fs::read(public_key_filename).unwrap();
                    let public_key = wcb.import_public_key_pem(&public_key_pem).unwrap();

                    let key = wcb.derive_key(private_key, public_key).unwrap();
                    let key_string = hex::encode(key);
                    println!("{}", key_string)
                }
                "encrypt" => {
                    let key_hex = &args[2];
                    let iv_hex = &args[3];

                    let key = hex::decode(key_hex).unwrap();
                    let iv = hex::decode(iv_hex).unwrap();

                    let mut data = Vec::new();
                    io::stdin().read_to_end(&mut data).unwrap();

                    let encrypted_data = wcb.encrypt(&key, &iv, &data).unwrap();
                    let encrypted_data_base64 = base64::encode(encrypted_data);

                    println!("{}", encrypted_data_base64)
                }
                "decrypt" => {
                    let key_hex = &args[2];
                    let iv_hex = &args[3];

                    let key = hex::decode(key_hex).unwrap();
                    let iv = hex::decode(iv_hex).unwrap();

                    let mut data_base64 = Vec::new();
                    io::stdin().read_to_end(&mut data_base64).unwrap();
                    let data = base64::decode(data_base64).unwrap();

                    let message = wcb.decrypt(&key, &iv, &data).unwrap();
                    let message_string = std::str::from_utf8(&message).unwrap();

                    println!("{}", message_string)
                }
                _ => help(),
            }
        }
        5 => {
            let cmd = &args[1];
            match &cmd[..] {
                "derive-and-encrypt" => {
                    let private_key_filename = &args[2];
                    let public_key_filename = &args[3];
                    let iv_hex = &args[4];

                    let mut data = Vec::new();
                    io::stdin().read_to_end(&mut data).unwrap();

                    let private_key_pem = fs::read(private_key_filename).unwrap();
                    let private_key = wcb.import_private_key_pem(&private_key_pem).unwrap();
                    let public_key_pem = fs::read(public_key_filename).unwrap();
                    let public_key = wcb.import_public_key_pem(&public_key_pem).unwrap();
                    let key = wcb.derive_key(private_key, public_key).unwrap();
                    let iv = hex::decode(iv_hex).unwrap();
                    let encrypted_data = wcb.encrypt(&key, &iv, &data).unwrap();
                    let encrypted_data_base64 = base64::encode(encrypted_data);

                    println!("{}", encrypted_data_base64)
                }
                "derive-and-decrypt" => {
                    let private_key_filename = &args[2];
                    let public_key_filename = &args[3];
                    let iv_hex = &args[4];
                    let mut data_base64 = Vec::new();
                    io::stdin().read_to_end(&mut data_base64).unwrap();

                    let private_key_pem = fs::read(private_key_filename).unwrap();
                    let private_key = wcb.import_private_key_pem(&private_key_pem).unwrap();
                    let public_key_pem = fs::read(public_key_filename).unwrap();
                    let public_key = wcb.import_public_key_pem(&public_key_pem).unwrap();
                    let key = wcb.derive_key(private_key, public_key).unwrap();
                    let iv = hex::decode(iv_hex).unwrap();
                    let data = base64::decode(data_base64).unwrap();
                    let message = wcb.decrypt(&key, &iv, &data).unwrap();
                    let message_string = std::str::from_utf8(&message).unwrap();

                    println!("{}", message_string)
                }
                _ => help(),
            }
        }
        _ => help(),
    }
}
